app/models/concerns/orchestration/dhcp.rb

```diff
@@ -78,13 +78,8 @@ def boot_server
     # if we don't manage tftp for IPv4 at all, we dont create a next-server entry.
     return unless tftp?
 
-    # first try to ask our IPv4 TFTP server for its boot server
-    bs = tftp.bootServer
-    # if that failed, trying to guess out tftp next server based on the smart proxy hostname
-    bs ||= URI.parse(subnet.tftp.url).host
-    # now convert it into an ip address (see http://theforeman.org/issues/show/1381)
-    ip = NicIpResolver.new(:nic => self).to_ip_address(bs) if bs.present?
```

Comment:
One more thing before I continue - there is this helper class called `NicIpResolver` which performs one quite confusing thing IMHO. It tries to query nameservers associated with the NIC-Subnet and if that fails it tries OS-level query via Ruby. This can be confusing when Subnet is misconfigured or entry is missing on the managed DNS server. Two options: A) I keep this behavior. B) I only will use OS-level.

Option (A) is more tricky because I am actually moving this out of our model context where NIC and Subnet is available to Proxy APi context where it's used from two places.

That's not all, there is also setting named "query_local_nameservers" and when set Foreman is supposed to use OS-level resolver instead reaching out to managed ones. However this code which I am trying to refactor completely ignores that setting.

If I want to do (B) or even (B) + utilizing the setting, then it's a bit off a problem when this is called from the ProxyStatus because there is no host context, therefore Subnet (thus managed DNS servers) is not known. If we want to fix the original bug report (status page) then I think option (A) is the only one but this creates confusion around the setting.

@ekohl and @ares

Reply:
What a mess! So the `NicIpResolver` actually:

* tries to resolve against authorized DNS servers defined in the domain
* then tries to resolve against:
 * against authorized DNS servers defined in the domain when query_local_nameservers is false
 * against OS DNS when query_local_nameservers is true

I am starting to think this needs a deeper redesign, probably we can introduce new Subnet flag TFTP server IPv4 address stored in Foreman database and some migration that will help users to fill it (proxy servers would need to be up during the upgrade or DNS correctly set). WHAT A MESS.

Reply:
I wonder if we should simplify this. Reading https://tools.ietf.org/html/rfc2132#section-9.4 nothing indicates this needs to be an IP but can actually be a DNS name. The ISC DHCP manual also suggests it's a text rather than an IP, indicating to me it can use DNS. So why do we need to do this translation at all? Perhaps this is a good example of garbage in, garbage out.

I'd appreciate it if you could try this out.

Reply:
From ISC DHCP man page: _The next-server statement is used to specify the host address of the server  from  which  the  initial  boot  file      (specified  in  the  filename  statement) is to be loaded.  Server-name should be a numeric IP address or a domain      name.  If no next-server statement applies to a given client, the address 0.0.0.0 is used._

However in the lease file next-server is stored as an IP with four hex octets. I'd need to test this, maybe some changes are required on the smart proxy level.

Anyway, how you like my idea to ditch /serverName call which never made sense to me and introducing new Subnet field called "TFTP server name" which could be either IPv4 or a hostname.

Reply:
> However in the lease file next-server is stored as an IP with four hex octets. I'd need to test this, maybe some changes are required on the smart proxy level.

Then we'd leave it up to the DHCP server implementation. Looks like our omapi implementation doesn't care about an IP. You're right that our lease file parser could break on this.

> Anyway, how you like my idea to ditch /serverName call which never made sense to me and introducing new Subnet field called "TFTP server name" which could be either IPv4 or a hostname.

Since we introduced capabilities, we also expose the setting. This means you can do `smart_proxy.setting('tftp', 'tftp_servername')`. If it registered correctly, then you have the value in your local database. We actually don't make use of this now.

Moving it to the Subnet model might make sense since it's more flexible. On the other hand, the current model of having the proxy export it actually gives you service discovery. You expose a TFTP service and that service knows how it's reachable (since you configured it).

Reply:
> Since we introduced capabilities, we also expose the setting. This means you can do smart_proxy.setting('tftp', 'tftp_servername').

Sure but why this setting is on the smart proxy in the first place? Smart proxy don't need to know this, it's the Foreman what uses this bit of information when sending DHCP create request. I believe this setting should be completely removed from proxy. The only purpose of this setting is to be returned via the API, why?

```
[lzap@box smart-proxy]$ ag tftp_servername
config/settings.d/tftp.yml
5:#:tftp_servername: tftp.domain.com
9::tftp_servername: 192.168.100.1

config/settings.d/tftp.yml.example
7:#:tftp_servername: tftp.domain.com

modules/tftp/tftp_plugin.rb
13:    expose_setting :tftp_servername

modules/tftp/tftp_api.rb
70:       {"serverName" => (Proxy::TFTP::Plugin.settings.tftp_servername || "")}.to_json

test/tftp/integration_test.rb
15:    Proxy::DefaultModuleLoader.any_instance.expects(:load_configuration_file).with('tftp.yml').returns(enabled: true, tftproot: '/var/lib/tftpboot', tftp_servername: 'tftp.example.com')
26:    expected_settings = { 'tftp_servername' => 'tftp.example.com' }

test/tftp/tftp_api_test.rb
115:    Proxy::TFTP::Plugin.settings.stubs(:tftp_servername).returns("servername")

test/migrations/migration_settings.yml
18::tftp_servername: 192.168.122.1

test/migrations/monolithic_config_file_migration_test.rb
35:                 :tftp_servername => "192.168.122.1"
```

Reply:
Well, lemme ask https://projects.theforeman.org/issues/800 :-)

Reply:
The purpose is service discovery. Let's say you have the following setup:
* foreman.example.com
* dhcp.example.com
* tftp.example.com

Now Foreman can query the foreman-proxy on tftp.example.com which address clients should use. Then on dhcp.example.com it can create the DHCP record with the correct TFTP server address.

If you didn't have this, then the user needs to manually do this. Since multiple DHCP servers can share the same TFTP proxy this is actually useful. All the user has to do is select the correct proxy for TFTP.

Reply:
Because if I am right, we can simply remove this completely, create Subnet.next-server DB field and warn user to either use IPv4 or hostname (but have correct DNS entry). And test this end to end if there are no changes required in our DHCP providers (I am little bit worried that some providers might expect IPv4 only). But that's it.

Reply:
You are right, let's simply improve the DNS resolution in this PR.